Add GOMAXPROCS and GOMEMLIMIT env vars to alloy container in the Helm chart

aerfio opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/alloy/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

- Introduce `GOMAXPROCS` env var to alloy container, which is responsible for limiting the number of OS threads that Go runtime will create, leading to better performance if Go app is limited by k8s limits. It's best practice for this env to match Linux container CPU quota. See the description of https://github.com/uber-go/automaxprocs for details
- Introduce `GOMEMLIMIT` env var to alloy container, which is soft runtime memory limit. More info: https://tip.golang.org/doc/gc-guide
- Remove useless dirs from packaged alloy chart

#### Which issue(s) this PR fixes

<!-- Uncomment the following line if you want that GitHub issue gets automatically closed after merging the PR -->
<!-- Fixes #issue_id -->

#### Notes to the Reviewer

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [ ] Documentation added
- [x] Tests updated
- [ ] Config converters updated

tpaschalis commented:
While this seems interesting, this is already something that people can add with extraEnv in their values.yaml, right?

Before making this the default for _all_ users, I'd personally like to experiment in Alloy instances of different sizes (really small to really huge) and make sure we're not seeing any real regression.

aerfio commented:
> While this seems interesting, this is already something that people can add with extraEnv in their values.yaml

Yeah, it's more about proper out of the box defaults. I 

> Before making this the default for all users, I'd personally like to experiment in Alloy instances of different sizes (really small to really huge) and make sure we're not seeing any real regression.

Sure, you can treat this PR as an invitation to experiment with that :D From my own experience and from what I've read online the best setting for `GOMAXPROCS` would be exactly the same as in this PR (see linked automaxprocs lib description), and as for `GOMEMLIMIT` I've read somewhere (I cant remember where exactly) to set it to ~90% of limits.memory, to give it a bit of space to overshoot this value without OOMKill, although some projects are also settting it to exactly limits.memory, for example [Crossplane](https://github.com/crossplane/crossplane/blob/b8a6fb90e0684adde826a6b647bb666d30395e9f/cluster/charts/crossplane/templates/deployment.yaml#L75-L86)

aerfio commented:
I've just had a situation in which GOMEMLIMIT helped me a lot. Due to change in our config (change from scraping logs using [loki.source.kubernetes](https://grafana.com/docs/agent/latest/flow/reference/components/loki.source.kubernetes/) to [loki.source.file](https://grafana.com/docs/agent/latest/flow/reference/components/loki.source.file/) we've seen a bunch of OOMs on 1 particular cluster (grafana-agent probably tried to process all of the logs on the node?). So my first thought was to try to use [stage.drop of loki.process](https://grafana.com/docs/agent/latest/flow/reference/components/loki.process/#stagedrop-block) and set it to small number and slowly increase it, but as an experiment I added GOMEMLIMIT and GOMAXPROCS, as in this PR. Grafana Agent pods stopped getting killed and they successfully processed those logs:

In this dashboard you can see mem usage before the change from loki.source.k8s (and a bunch of other things), then loki.source.file, then I added GOMEMLIMIT and GOMAXPROCS

<img width="2493" alt="image" src="https://github.com/grafana/alloy/assets/17271979/0a73f15b-2ae5-4689-b1aa-a0da614147ad">

<img width="2341" alt="image" src="https://github.com/grafana/alloy/assets/17271979/f80c025e-3908-4978-8f07-a94b4506f2bf">

It also might have been a bit of luck that in this precise moment that I added GOMEMLIMIT our pods stopped getting OOMKilled, you probably have much bigger deployments in Grafana then I do, measure it yourself 👍🏻 I'm happy to close this PR if your tests prove that these env vars are not helpful at all in your case and just add it in my values.yaml

tpaschalis commented:
cc @mattdurham who was also planning to experiment with these values.

aerfio commented:
I see that https://github.com/grafana/alloy/pull/655 was merged so GOMEMLIMIT is taken care of, should I close this PR @tpaschalis ? I think you can also take care of GOMAXPROCS internally (or not, but I also saw this env used internally in chart of mimir and tempo (iirc))

aerfio commented:
Closing this as it's already partly done and Grafana team will probably take this on by themselves
